File: src/Homeworks/Homework1.Test/ParserTests.cs
using System.Collections.Generic;
using Xunit;

namespace Homework1.Test
{
    public class ParserTests
    {
        [Fact]
        public void TryParseArguments_CorrectExpression_ReturnZero()
        {
            var actual = Parser.TryParseArguments(
                new[] {"1", "+", "2"}, out _, out _, out _);
            Assert.Equal(0, actual);
        }

        [Theory]
        [MemberData(nameof(InvalidArguments))]
        public void TryParseArguments_InvalidArguments_ReturnOne(string[] args)
        {
            var actual = Parser.TryParseArguments(
                args, out _, out _, out _);
            Assert.Equal(1, actual);
        }

        public static IEnumerable<object[]> InvalidArguments()
        {
            yield return new object[] {new[]{"x", "-", "2"}};
            yield return new object[] {new[]{"5", "+", "="}};
            yield return new object[] {new[]{"p", "+", "x"}};
        }
        
        [Theory]
        [MemberData(nameof(InvalidOperations))]
        public void TryParseArguments_InvalidOperation_ReturnTwo(string[] args)
        {
            var actual = Parser.TryParseArguments(
                args, out _, out _, out _);
            Assert.Equal(2, actual);
        }

        public static IEnumerable<object[]> InvalidOperations()
        {
            yield return new object[] {new[]{"10", "<", "2"}};
            yield return new object[] {new[]{"5", "\\", "5"}};
            yield return new object[] {new[]{"10", "%", "6"}};
        }
    }
}

File: src/Homeworks/Homework5.Tests/ParserTests.fs
module Tests.Parser

open Homework5
open Xunit

let CorrectArguments = [|
    [|[|"10"; "+"; "2"|]|]
    [|[|"2"; "-"; "10"|]|]
    [|[|"20"; "*"; "10.23"|]|]
    [|[|"4.12231"; "/"; "4"|]|]
    [|[|"50.124435"; ":"; "2.21"|]|]
|]

[<Theory>]
[<MemberData(nameof(CorrectArguments))>]
let TryParseArguments_CorrectArguments_ReturnOk args =
    let actual = Parser.tryParseArguments args
    match actual with
    | Ok r -> Assert.Equal((args.[0] |> decimal,args.[2] |> decimal), r)
    | Error e -> Assert.True(false)
    
let InvalidArguments = [|
    [|[|"x"; "-"; "2"|]|]
    [|[|"5.19"; "+"; "="|]|]
    [|[|"p"; "+"; "x"|]|]
|]

[<Theory>]
[<MemberData(nameof(InvalidArguments))>]
let TryParseArguments_InvalidArguments_ReturnError args =
    let actual = Parser.tryParseArguments args
    match actual with
    | Ok r -> Assert.True(false)
    | Error e -> Assert.Equal(1, e)
    
let InvalidOperations = [|
    [|[|"10"; "<"; "2"|]|]
    [|[|"5"; "\\"; "5"|]|]
    [|[|"10"; "%"; "6"|]|]
|]

[<Theory>]
[<MemberData(nameof(InvalidOperations))>]
let TryParseOperation_InvalidOperation_ReturnError args =
    let actual = Parser.tryParseOperation args
    match actual with
    | Ok r -> Assert.True(false)
    | Error e -> Assert.Equal(2, e)

let CorrectOperations : obj[][] = [|
    [|[|"10"; "+"; "2"|]; Plus|]
    [|[|"2"; "-"; "10"|]; Minus|]
    [|[|"20"; "*"; "10"|]; Multiply|]
    [|[|"4"; "/"; "4"|]; Divide|]
    [|[|"50"; ":"; "2"|]; Divide|]
|]

[<Theory>]
[<MemberData(nameof(CorrectOperations))>]
let tryParseOperation_CorrectOperations_ReturnOk args exp =
    let actual = Parser.tryParseOperation args
    match actual with
    | Ok o -> Assert.Equal(exp, o)
    | Error e -> Assert.True(false)